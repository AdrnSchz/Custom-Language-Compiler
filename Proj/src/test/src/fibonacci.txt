<< Program to compute fibonacci sequence
    of a number
	Author: Marc
	Version: 3.33
	Date: 7-5-2023 >>

<< Globals declaration >>
fact bro ZERO = 0.
fact bro ONE = 1.

<< Fibonacci sequence >>
bro fibonacci ¿ bro num ? ¡
 	swipe ¿ num ? ¡
     	right 0:
  			throwback ZERO.
  		right 1:
 		    throwback ONE.
 		left:
 		    bro fVal = fibonacci ¿num - 1?.
 		    fVal += fibonacci ¿num - 2?.
 			throwback fVal.
 	!
 	throwback ZERO.
 !

<< Main function of the program
	Author: Adrian
	Version: 3.33
	Date: 2-3-2023 >>
CEO ¡
    bro num = 6.
    << Compute fibonacci >>
	bro fib = fibonacci¿ num ?.
!