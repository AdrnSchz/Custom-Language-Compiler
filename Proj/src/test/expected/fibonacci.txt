	.data
TRUE: .float   1
	.text
	.data
FALSE: .float   0
	.text
	li   $fp, 2147483644
	move $sp, $fp
	.data
ZERO_: .word   0
	.text
	li   $t0, 0
	sub  $sp, $sp, 4
	lw   $t1, ZERO_
	move $t1, $t0
	.data
ONE_: .word   0
	.text
	li   $t2, 1
	sub  $sp, $sp, 4
	lw   $t3, ONE_
	move $t3, $t2
	sw   $t0, -4($fp)
	sw   $t3, ONE_
	sw   $t2, -8($fp)
	sw   $t1, ZERO_
	j    _CEO
_fibonacci:
	sw   $ra, 0($fp)
	j    check0
if0:
	lw   $t4, ZERO_
	move $v0, $t4
	move $sp, $fp
	addi $sp, $sp, 8
	lw   $ra, 0($fp)
	lw   $fp, 4($fp)
	jr   $ra
	j    continue0
elif1:
	lw   $t5, ONE_
	move $v0, $t5
	move $sp, $fp
	addi $sp, $sp, 8
	lw   $ra, 0($fp)
	lw   $fp, 4($fp)
	jr   $ra
	j    continue0
else2:
	sub  $sp, $sp, 4
	li   $t6, 1
	sub  $sp, $sp, 4
	lw   $t8, -4($fp)
	sub  $t9, $t8, $t6
	move $t7, $t9
	sw   $t6, -12($fp)
	sw   $t8, -4($fp)
	sw   $fp, -4($sp)
	sub  $sp, $sp, 8
	move $fp, $sp
	sw   $t7, -4($sp)
	sub  $sp, $sp, 4
	jal  _fibonacci
	lw   $t9, -8($fp)
	move $t9, $v0
	li   $t0, 2
	sub  $sp, $sp, 4
	lw   $t2, -4($fp)
	sub  $t3, $t2, $t0
	move $t1, $t3
	sw   $t9, -8($fp)
	sw   $t2, -4($fp)
	sw   $t0, -16($fp)
	sw   $fp, -4($sp)
	sub  $sp, $sp, 8
	move $fp, $sp
	sw   $t1, -4($sp)
	sub  $sp, $sp, 4
	jal  _fibonacci
	lw   $t4, -8($fp)
	add  $t5, $t4, $v0
	move $t3, $t5
	move $t4, $t3
	move $v0, $t4
	move $sp, $fp
	addi $sp, $sp, 8
	lw   $ra, 0($fp)
	lw   $fp, 4($fp)
	jr   $ra
	j    continue0
check0:
	li   $t5, 0
	sub  $sp, $sp, 4
	lw   $t7, -4($fp)
	mtc1 $t5, $f0
	cvt.s.w $f0, $f0
	mtc1 $t7, $f2
	cvt.s.w $f2, $f2
	c.eq.s $f0, $f2
	l.s   $f0, FALSE
	l.s   $f2, TRUE
	movt.s $f0, $f2
	cvt.w.s $f0, $f0
	mfc1 $t6, $f0
	li   $t8, 1
	sub  $sp, $sp, 4
	mtc1 $t8, $f2
	cvt.s.w $f2, $f2
	mtc1 $t7, $f4
	cvt.s.w $f4, $f4
	c.eq.s $f2, $f4
	l.s   $f2, FALSE
	l.s   $f4, TRUE
	movt.s $f2, $f4
	cvt.w.s $f2, $f2
	mfc1 $t9, $f2
	bgtz $t6, if0
	bgtz $t9, elif1
	j    else2
continue0:
	lw   $t0, ZERO_
	move $v0, $t0
	move $sp, $fp
	addi $sp, $sp, 8
	lw   $ra, 0($fp)
	lw   $fp, 4($fp)
	jr   $ra
_CEO:
	sub  $sp, $sp, 4
	li   $t1, 6
	sub  $sp, $sp, 4
	lw   $t2, -4($fp)
	move $t2, $t1
	sub  $sp, $sp, 4
	sw   $t2, -4($fp)
	sw   $t1, -8($fp)
	sw   $fp, -4($sp)
	sub  $sp, $sp, 8
	move $fp, $sp
	sw   $t2, -4($sp)
	sub  $sp, $sp, 4
	jal  _fibonacci
	lw   $t3, -12($fp)
	move $t3, $v0
	li   $v0, 10
	syscall
